### ISSUE TITLE
false positive: useless_format when inside a macro

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang-nursery/rust-clippy#2424]
The `useless_format` lint has a false positive when used inside a macro, see it [live](https://play.rust-lang.org/?gist=4237515f15c4dd313195f6831c29892b&version=stable):

```rust
struct Foo(pub String);

macro_rules! foo {
  ($($t:tt)*) => (Foo(format!($($t)*)))
}

fn main() {
    let _a = foo!("what's {}", "up"); // OK
    let _b = foo!("a false positive"); // WRONG
}
```

produces

```shell
warning: useless use of `format!`
 --> src/main.rs:4:23
  |
4 |   ($($t:tt)*) => (Foo(format!($($t)*)))
  |                       ^^^^^^^^^^^^^^^
...
9 |     let _b = foo!("a false positive"); // WRONG
  |              ------------------------ in this macro invocation
  |
  = note: #[warn(useless_format)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.184/index.html#useless_format
```

So.. yeah, that use of `format!` is useless, but if the macro is outside your control there is nothing one can do about it.

**Comment by killercup** [comment]
dupe of #1130?
